lia-ui/src/main/java/com/lithium/community/android/ui/components/adapters/LiConversationAdapter.java

```diff
@@ -169,136 +185,318 @@ private void showPopupMenu(final View view, final int position) {
             popup.getMenu().removeItem(R.id.li_action_delete);
         }
 
-        popup.setOnMenuItemClickListener(new PopupMenu.OnMenuItemClickListener() {
-            @Override
-            public boolean onMenuItemClick(MenuItem menuItem) {
-                if (menuItem.getItemId() == R.id.li_action_report_abuse) {
-                    Bundle bundle = new Bundle();
-                    bundle.putString(LiSDKConstants.REPORT_ABUSE_MESSAGE_ID, item.getId() + "");
-                    bundle.putString(LiSDKConstants.REPORT_ABUSE_AUTHOR_ID, item.getAuthor().getId() + "");
-                    Intent i = new Intent(activity, LiReportAbuseActivity.class);
-                    i.putExtras(bundle);
-                    activity.startActivity(i);
-                    return true;
-                } else if (menuItem.getItemId() == R.id.li_action_mark_read) {
-                    if (item.getUserContext() != null) {
-                        boolean isMarkUnread = false;
-                        if (item.getUserContext() != null && item.getUserContext().getRead()) {
-                            isMarkUnread = true;
+        boolean showSubscribe = (position == 0);
+        boolean isSubscribed = item.getUserContext().isSubscribed();
+        popup.getMenu().findItem(R.id.li_action_subscribe).setVisible(showSubscribe);
+        popup.getMenu().findItem(R.id.li_action_subscribe).setTitle(isSubscribed ? R.string.li_unsubscribe : R.string.li_subscribe);
+        popup.setOnMenuItemClickListener(new PopupMenuClickListener(item, getItems().size() == 1 || position == 0, position));
+        popup.show();
+    }
+
+    private void actionInProgress(boolean inProgress) {
+        if (mActionInProgressListener != null) {
+            mActionInProgressListener.actionInProgress(inProgress);
+        }
+    }
+
+    private class PopupMenuClickListener implements PopupMenu.OnMenuItemClickListener {
+
+        private final LiMessage item;
+        private final boolean isParentItem;
+        private final int position;
+
+        PopupMenuClickListener(LiMessage item, boolean isParentItem, int position) {
+            this.isParentItem = isParentItem;
+            this.position = position;
+            this.item = item;
+        }
+
+        @Override
+        public boolean onMenuItemClick(MenuItem menuItem) {
+            if (menuItem.getItemId() == R.id.li_action_report_abuse) {
+                openReportAbuse(item);
+                return true;
+            } else if (menuItem.getItemId() == R.id.li_action_mark_read) {
+                markRead(item);
+                return true;
+            } else if (menuItem.getItemId() == R.id.li_action_delete) {
+
+                DialogInterface.OnClickListener dialogClickListener = new DialogInterface.OnClickListener() {
+                    @Override
+                    public void onClick(final DialogInterface dialog, int which) {
```

Comment:
please use appropriate param/var names (which??)

Reply:
Also both params should be final

Reply:
No need, it is auto generated by SDK, in android which means one of the cases ( positive, negative & neutral)